log-warper-core/src/Log/Core/Class.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE DefaultSignatures          #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | Monad type class for things that have logging ability.
+
+module Log.Core.Class
+       ( -- * Interface
+         MonadLogger (..)
+       , logR
+       , withR
+       , unliftLogAction
+
+         -- * Basic interface implementation
+       , ActionT (..)
+       , usingActionT
+       ) where
+
+import Control.Monad (void)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Class (MonadTrans (..))
+import Control.Monad.Trans.Except (ExceptT, mapExceptT, runExceptT)
+import Control.Monad.Trans.Identity (IdentityT (..), mapIdentityT)
+import Control.Monad.Trans.Maybe (MaybeT (..), mapMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask, mapReaderT)
+import Control.Monad.Trans.State.Strict (StateT (..), get, mapStateT)
+import Data.Kind (Type)
+
+import Log.Core.Action (LogAction (..), cmap)
+
+{- | Type class for monads that have logging ability. In order to log things
+monad @m@ should be able to construct 'LogAction' for some logging record @r@.
+
+__Laws:__
+
+Some laws duplicate @lens@ laws. And maybe of the laws are redundant and can be
```

Comment:
"maybe of the laws"? Did you skip a word?

Reply:
@int-index Yes, I did. This should be "maybe some of the laws".